src/compiler/parser.ts

```diff
@@ -3569,30 +3569,60 @@ namespace ts {
             return parsePostfixTypeOrHigher();
         }
 
+        function parseFunctionOrConstructorTypeToError(
+            functionTypeDiagnostic: DiagnosticMessage,
+            constructorTypeDiagnostic: DiagnosticMessage

```

Comment:
Instead of passing along the diagnostics, I might just pass along the information of whether or not this is a union type or an intersection type.

Reply:
@DanielRosenwasser Thank you for your reviews!
I updated this function so it new receives a single parameter `isInUnionType: boolean`.